api/tests/DataPersister/ContentNodes/ColumnLayoutDataPersisterTest.php

```diff
@@ -32,60 +33,31 @@ protected function setUp(): void {
         $this->contentNode->prototype->contentType = new ContentType();
         $this->contentNode->prototype->contentType->name = 'test';
 
-        $this->dataPersister = new ColumnLayoutDataPersister($this->decoratedMock);
-    }
-
-    public function testDelegatesSupportCheckToDecorated() {
-        $this->decoratedMock
-            ->expects($this->exactly(2))
-            ->method('supports')
-            ->willReturnOnConsecutiveCalls(true, false)
-        ;
-
-        $this->assertTrue($this->dataPersister->supports($this->contentNode, []));
-        $this->assertFalse($this->dataPersister->supports($this->contentNode, []));
+        $this->dataPersister = new ColumnLayoutDataPersister($this->dataPersisterObservable);
     }
 
     public function testDoesNotSupportNonColumnLayout() {
-        $this->decoratedMock
+        $this->dataPersisterObservable
             ->method('supports')
             ->willReturn(true)
         ;
 
         $this->assertFalse($this->dataPersister->supports([], []));
     }
 
-    public function testDelegatesPersistToDecorated() {
-        // given
-        $this->decoratedMock->expects($this->once())
-            ->method('persist')
-        ;
-
-        // when
-        $this->dataPersister->persist($this->contentNode, []);
-
-        // then
-    }
```

Comment:
Ist dieser Test wirklich nicht mehr relevant? Es ist ja weiterhin wichtig dass man den `parent` aufruft und nicht aus Versehen `return null;` hinschreibt

Reply:
Ist halt Logik aus dem `AbstractDataPersister`. Ist jetzt schon die Frage, ob man diese Logik in jedem einzelnen Data Persister Test wiederholen will, oder ob wir nicht besser die Klasse `AbstractDataPersister` ausgiebig testen wollen.

Reply:
Ja, dass `$decorated->persist` aufgerufen wird ist schon Logik aus dem `AbstractDataPersister`. Aber was ich meine ist, fehlt nicht ein Test der den Unterschied zwischen folgenden beiden Implementationen überprüft:
```php
public function beforeCreate($data): ColumnLayout {
    if (isset($data->prototype)) {
        // do something
    }

    return parent::beforeCreate($data);
}
```
und
```php
public function beforeCreate($data): ColumnLayout {
    if (isset($data->prototype)) {
        // do something
    }

    return $data;
}
```

Das ist ein Fehler der mir auch schon passiert ist. Es kann auch je nach Fall mal das eine oder das andere richtig sein. Und das kann man nicht im `AbstractDataPersister` testen.

Reply:
Das wird in `testCopyFromPrototypeOnCreate` geprüft. Dort wird getetet, dass `instanceName`, `slot`, `position` und `contentType` vom prototype kopiert werden, was wiederum im `parent::beforeCreate` (=`ContentNodeAbstractDataPersister::beforeCreate`) passiert.

Ab hier:
https://github.com/ecamp/ecamp3/pull/2195/files#diff-d780655bb754c71a7d074aeed10174ae1ba516512c4c18c629bbe19989611df5R67

Reply:
Hm, nevermind, was ich mir wünsche geht wies aussieht gar nicht. Ich kann ja nicht den `parent::beforeCreate` Call mocken, und beim `ContentNodeAbstractDataPersister` ist die einzige Logik dort drin auch bereits in deinen Tests abgedeckt.
Genauer gesagt testen deine Tests ausschliesslich Logik vom ContentNodeAbstractDataPersister. Ich kann die Logik aus ColumnLayoutDataPersister komplett löschen, und solange der `parent` Call drin ist passen die Tests.

Reply:
Aha, lustig, das ist ein Seiteneffekt der `getColumns` Methode die die Columns des Prototypes zurückgibt wenn auf dem Objekt keiner gesetzt ist. Ich mach dir einen PR in deinen PR als Vorschlag wie ich damit umgehen würde.

Reply:
Voilà: https://github.com/usu/ecamp3/pull/3

Ich mach dann mal noch einen PR mit Validierung dass der Prototype vom gleichen Typ ist wie der ContentNode. Dann sind all die nackten Typecheck-`Exception`s in der Praxis dann nicht mehr möglich.

Reply:
> Aha, lustig, das ist ein Seiteneffekt der `getColumns` Methode die die Columns des Prototypes zurückgibt wenn auf dem Objekt keiner gesetzt ist. Ich mach dir einen PR in deinen PR als Vorschlag wie ich damit umgehen würde.

Yep, genau ins gleiche hineingelaufen :-) Aber dein Vorschlag gefällt mir. Ist in den Branch gemerged.

Reply:
Sorry, die Tests sind noch kaputt weil ich ein Example abgeändert habe. Fix folgt direkt.